compiler/optimizer/op.go

```diff
@@ -2,121 +2,120 @@ package optimizer
 			}
 		}
-		return sortKey, nil
+		return out, nil
 	case *dag.Summarize:
-		return analyzeOpSummarize(op, sortKey), nil
+		if isKeyOfSummarize(op, in) {
+			return in, nil
+		}
+		return order.Nil, nil
 	case *dag.Put:
 		for _, assignment := range op.Args {
 			if fieldOf(assignment.LHS).Equal(key) {
 				return order.Nil, nil
 			}
 		}
-		return sortKey, nil
+		return in, nil
 	case *dag.Sequential:
+		out := in
 		for _, op := range op.Ops {
 			var err error
-			sortKey, err = o.analyzeOp(op, sortKey)
+			out, err = o.analyzeSortKey(op, out)
 			if err != nil {
 				return order.Nil, err
 			}
 		}
-		return sortKey, nil
+		return out, nil
 	case *dag.Sort:
-		// XXX Only single sort keys.  See issue #2657.
-		if len(op.Args) != 1 {
-			return order.Nil, nil
-		}
-		newKey := fieldOf(op.Args[0])
-		if newKey == nil {
-			// Not a field
-			return order.Nil, nil
-		}
-		return order.NewSortKey(op.Order, field.List{key}), nil
-	case *dag.From:
-		var egress order.SortKey
-		for k := range op.Trunks {
-			trunk := &op.Trunks[k]
-			l, err := o.sortKeyOfSource(trunk.Source, sortKey)
-			if err != nil || l.IsNil() {
-				return order.Nil, err
-			}
-			l, err = o.analyzeOp(trunk.Seq, l)
-			if err != nil {
-				return order.Nil, err
-			}
-			if k == 0 {
-				egress = l
-			} else if !egress.Equal(l) {
-				return order.Nil, nil
-			}
-		}
-		return egress, nil
+		return sortKeyOfSort(op), nil
 	default:
 		return order.Nil, nil
 	}
 }
 
-// summarizeOrderAndAssign determines whether its first groupby key is the
-// same as the scan order or an order-preserving function thereof, and if so,
-// sets ast.Summarize.InputSortDir to the propagated scan order.  It returns
-// the new order (or order.Nil if unknown) that will arise after the summarize
-// is applied to its input.
-func analyzeOpSummarize(summarize *dag.Summarize, sortKey order.SortKey) order.SortKey {
-	// Set p.InputSortDir and return true if the first grouping key
-	// is inputSortField or an order-preserving function of it.
-	key := sortKey.Keys[0]
-	if len(summarize.Keys) == 0 {
+func sortKeyOfSort(op *dag.Sort) order.SortKey {
+	// XXX Only single sort keys.  See issue #2657.
+	if len(op.Args) != 1 {
 		return order.Nil
 	}
-	groupByKey := fieldOf(summarize.Keys[0].LHS)
-	if groupByKey.Equal(key) {
-		rhsExpr := summarize.Keys[0].RHS
-		rhs := fieldOf(rhsExpr)
-		if rhs.Equal(key) || orderPreservingCall(rhsExpr, groupByKey) {
-			return sortKey
+	return sortKeyOfExpr(op.Args[0], op.Order)
+}
+
+func sortKeyOfExpr(e dag.Expr, o order.Which) order.SortKey {
+	key := fieldOf(e)
+	if key == nil {
+		return order.Nil
+	}
+	return order.NewSortKey(o, field.List{key})
+}
+
+// isKeyOfSummarize returns true iff its first groupby key is the
+// same as the given sort order or an order-preserving function thereof.
+func isKeyOfSummarize(summarize *dag.Summarize, in order.SortKey) bool {
+	key := in.Keys[0]
+	for _, outputKeyExpr := range summarize.Keys {
```

Comment:
My understanding is that we're only interested in whether or not the first summarize key (`summarize.Keys[0]`) matches `in`, so why examine anything past that first key?

Reply:
The comment is wrong.  I'll fix it.  We want to know if any group-by key is corresponds to sort order so we can release group-by results as soon as its known that there won't be anymore for a given value.

Reply:
But runtime/op/groupby/groupby.go only looks at the first summarize key (via Aggregator.keyCompare) to make that decision. Given that, is there any reason to look beyond the first key here?